### DISCUSSION TITLE
tsh rsync ?

### DISCUSSION THREAD

**Discussion by dfang** [gravitational/teleport#24902] [Q&A]
since there is `tsh scp` command for transferring files between nodes, why not rsync ?

**Comment by webvictim** [comment]
You can run `rsync -e 'tsh ssh'` to achieve the same thing.

**Reply by dfang** [reply]
thanks, but it's inconvenient to copy files between 2 nodes.

btw, `tsh scp` does not support coping files between two remotes like `scp -3` too.

**Reply by webvictim** [reply]
Good feature request :) https://github.com/gravitational/teleport/issues/25566

**Reply by sam-s** [reply]
> You can run `rsync -e 'tsh ssh'` to achieve the same thing.

alas, does not work on windows:
```
rsync -av -e '/c/progra~1/teleport/tsh.exe scp' ...
tsh: error: unknown long flag '--server'
usage: tsh scp [<flags>] <from, to>...
...
```

**Reply by webvictim** [reply]
Well, that's because you're running `tsh scp`, not `tsh ssh`.

`rsync -av -e '/c/progra~1/teleport/tsh.exe ssh' ...` should work.

**Reply by sam-s** [reply]
> Well, that's because you're running `tsh scp`, not `tsh ssh`.
> 
> `rsync -av -e '/c/progra~1/teleport/tsh.exe ssh' ...` should work.

Sorry about the typo.
```
$ rsync -av -e '/c/progra~1/teleport/tsh.exe ssh' ...
rsync: connection unexpectedly closed (0 bytes received so far) [Receiver]
rsync error: error in rsync protocol data stream (code 12) at io.c(226) [Receiver=3.1.3]
ERROR: Process exited with status 12

rsync: [sender] safe_read failed to read 4 bytes: Connection reset by peer (104)
rsync error: error in rsync protocol data stream (code 12) at io.c(282) [sender=3.3.0]
      0 [sig] rsync 2951! sigpacket::process: Suppressing signal 30 to win32 process (pid 34848)
      0 [sig] rsync 2951! sigpacket::process: Suppressing signal 30 to win32 process (pid 34848)
```

**Reply by webvictim** [reply]
What does `/c/progra~1/teleport/tsh.exe status` say?

**Reply by sam-s** [reply]
> What does `/c/progra~1/teleport/tsh.exe status` say?

```
$ tsh status
> Profile URL:        https://XXXXX:443
  Logged in as:       ZZZZ
  Cluster:            XXXXX
  Roles:              yyy
  Logins:             ZZZZ
  Kubernetes:         enabled
  Valid until:        2024-12-19 03:29:47 -0500 EST [valid for 11h23m0s]
  Extensions:         login-ip, permit-X11-forwarding, permit-agent-forwarding, permit-port-forwardi
ng, permit-pty, private-key-policy
```

**Reply by webvictim** [reply]
Yeah, in that case my guess is it's just not providing the data stream in the right format.

It looks like you're using Cygwin? You might have more joy with WSL and using actual `tsh` for Linux.

**Reply by sam-s** [reply]
Alas, WSL is not available to me.
I am using msys, not cygwin.
